File: OJ/leet986/leet986.cpp
/*
leet986


*/


#include <iostream>
#include <tuple>
#include <vector>
#include <queue>
#include <stack>
#include <string>
#include <set>
#include <map>
#include <unordered_set>
#include <unordered_map>
#include <algorithm>
#include <numeric>
#include <cassert>
#include <functional>
#include <utility>
#include "../utils/LeetCpp.utils.hpp"

using namespace std;

class Solution {
public:
    vector<vector<int>> intervalIntersection(const vector<vector<int>>& V1, const vector<vector<int>>& V2) {
        // O(M+N)
        vector<vector<int>> ans;
        auto p1 = V1.begin(), p2 = V2.begin();
        while (p1 != V1.end() and p2 != V2.end()) {
            auto lo = max((*p1)[0], (*p2)[0]);
            auto hi = min((*p1)[1], (*p2)[1]);
            if (lo <= hi) ans.push_back({lo, hi});
            if ((*p1)[1] < (*p2)[1]) p1++;
            else p2++;
        }
        return ans;
    }
};

int main(int argc, char const *argv[])
{
    Solution sol;   
    cout << sol.intervalIntersection({{0,2},{5,10},{13,23},{24,25}}, {{1,5},{8,12},{15,24},{25,26}}) << endl;
    return 0;
}

File: OJ/leet986/leet986.py
# leet986

from typing import List


class Solution:
    def intervalIntersection(self, firstList: List[List[int]], secondList: List[List[int]]) -> List[List[int]]:
        # 这个解法不是非常优雅
        ans = []
        i2, n2 = 0, len(secondList)
        for v1 in firstList:
            while i2 < n2:
                v2 = secondList[i2]
                if v2[1] < v1[0]: i2 += 1
                elif v2[0] > v1[1]: break 
                else:
                    ans.append([max(v1[0], v2[0]), min(v1[1], v2[1])])
                    i2 += 1
            if i2: i2 -= 1
        return ans  
            
            


print(Solution().intervalIntersection(firstList = [[0,2],[5,10],[13,23],[24,25]], secondList = [[1,5],[8,12],[15,24],[25,26]] ))
print(Solution().intervalIntersection(firstList = [[1,3],[5,9]], secondList = []))
print(Solution().intervalIntersection(firstList = [], secondList = [[4,8],[10,12]]))
print(Solution().intervalIntersection(firstList = [[1,7]], secondList = [[3,10]]))
print(Solution().intervalIntersection([[4,6],[7,8],[10,17]],[[5,10]]))


